### PR TITLE
#26 added new module units-viewbinding

### PR DESCRIPTION
**Pull request by anton6tak**

### CODE REVIEW

File: sample/android-viewbinding/src/main/java/com/icerockdev/MainActivity.kt [hunk 1]
```diff
@@ -79,12 +75,12 @@ class MainActivity : AppCompatActivity() {
             }
         }
         recyclerView.layoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)
-        recyclerView.adapter = UnitsRecyclerViewAdapter(this).apply {
-            units = test.getUnits()
-        }
+//        recyclerView.adapter = UnitsRecyclerViewAdapter(this).apply {
+//            units = test.getUnits()
+//        }
```
Comment (Alex009) [inline sample/android-viewbinding/src/main/java/com/icerockdev/MainActivity.kt]: uncomment this

File: build.gradle.kts [hunk 2]
```diff
@@ -2,6 +2,10 @@
  * Copyright 2019 IceRock MAG Inc. Use of this source code is governed by the Apache 2.0 license.
  */
 
+plugins {
+    id("org.jetbrains.kotlinx.binary-compatibility-validator") version "0.6.0"
```
Comment (Alex009) [inline build.gradle.kts:6]: why this added?

File: gradle/libs.versions.toml [hunk 3]
```diff
@@ -3,13 +3,16 @@ kotlinVersion = "1.5.20"
...
 androidCoreTestingVersion = "2.1.0"
+androidLifecycleVersion = "2.2.0"
 recyclerViewVersion = "1.2.1"
 kotlinPoetVersion = "1.3.0"
 coroutinesVersion = "1.5.1-native-mt"
 mokoGraphicsVersion = "0.8.0"
 mokoResourcesVersion = "0.16.2"
 mokoTestVersion = "0.4.0"
 mokoUnitsVersion = "0.6.2"
+mokoMvvmVersion = "0.11.0"
```
Comment (Alex009) [inline gradle/libs.versions.toml:14]: why mvvm added?

File: sample/android-viewbinding/src/main/java/com/icerockdev/BR.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  public static int _all = 0;
+
+  public static int number = 1;
+
+  public static int obj1 = 2;
+
+  public static int obj2 = 3;
+
+  public static int text = 4;
+}
```
Comment (Alex009) [inline sample/android-viewbinding/src/main/java/com/icerockdev/BR.java:13]: what is it? for what?

File: sample/android-viewbinding/src/main/java/com/icerockdev/UnitSimpleDropdown.kt [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+    override fun UnitSimpleDropdownBinding.bindData(
+        context: Context,
+        lifecycleOwner: LifecycleOwner,
+        viewHolder: VBViewHolder<UnitSimpleDropdownBinding>
+    ) {
+        unitSimpleDropdownTextView.text = text
+    }
+
+    class Combined(itemId: Long, text: String) : UnitSimple(itemId, text),
```
Comment (Alex009) [inline sample/android-viewbinding/src/main/java/com/icerockdev/UnitSimpleDropdown.kt:33]: this variant should be described in readme. and all about viewbinding usage should be in readme

File: units-viewbinding/src/main/kotlin/dev/icerock/moko/units/viewbinding/VBViewHolder.kt [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import java.io.Closeable
+
+class VBViewHolder<VB : ViewBinding>(
+    val binding: VB,
+    val lifecycleOwner: LifecycleOwner,
+) : RecyclerView.ViewHolder(binding.root) {
+    val context: Context get() = itemView.context
+
+    // maybe change to `associatedObject: Any? = null` ?
+    var closeable: Closeable? = null
```
Comment (Alex009) [inline units-viewbinding/src/main/kotlin/dev/icerock/moko/units/viewbinding/VBViewHolder.kt:20]: need to change to `storedRef`

File: units/build.gradle.kts [hunk 7]
```diff
@@ -18,6 +18,7 @@ android.buildFeatures.dataBinding = true
 dependencies {
     androidMainImplementation(libs.appCompat)
     androidMainApi(libs.recyclerView)
+    androidMainApi(libs.mokoMvvmLiveData)
```
Comment (Alex009) [inline units/build.gradle.kts:21]: why?

File: units/src/androidMain/kotlin/dev/icerock/moko/units/databinding/BindingAdapters.kt [hunk 8]
```diff
@@ -95,3 +99,40 @@ private fun getLifecycleOwnerFromContext(context: Context): LifecycleOwner {
     if (context is ContextWrapper) return getLifecycleOwnerFromContext(context.baseContext)
     throw IllegalArgumentException("context $context not implement LifecycleOwner")
 }
+
+fun RecyclerView.bindUnits(lifecycleOwner: LifecycleOwner, units: LiveData<List<UnitItem>>) {
```
Comment (Alex009) [inline units/src/androidMain/kotlin/dev/icerock/moko/units/databinding/BindingAdapters.kt:103]: all databinding logic should be extracted to `units-databinding` artifact. and need to remove livedata dependency

File: units-viewbinding/build.gradle.kts [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+plugins {
+    id("android-library-convention")
+    id("detekt-convention")
+    id("kotlin-kapt")
+    id("android-publication-convention")
+    id("kotlin-android")
+}
+
+group = "dev.icerock.moko"
+version = libs.versions.mokoUnitsVersion.get()
```
Comment (Alex009) [inline units-viewbinding/build.gradle.kts:14]: group and version not set in root build.gradle?

**Review by Alex009 — CHANGES_REQUESTED** [review]

File: sample/android-viewbinding/src/main/java/com/icerockdev/MainActivity.kt (hunk 1 above)
Comment (anton6tak) [reply sample/android-viewbinding/src/main/java/com/icerockdev/MainActivity.kt]: fixed

File: gradle/libs.versions.toml (hunk 3 above)
Comment (anton6tak) [reply gradle/libs.versions.toml:14]: removed

File: units/build.gradle.kts (hunk 7 above)
Comment (anton6tak) [reply units/build.gradle.kts:21]: removed

File: units-viewbinding/src/main/kotlin/dev/icerock/moko/units/viewbinding/VBViewHolder.kt (hunk 6 above)
Comment (anton6tak) [reply units-viewbinding/src/main/kotlin/dev/icerock/moko/units/viewbinding/VBViewHolder.kt:20]: changed

File: sample/mpp-library/build.gradle.kts [hunk 10]
```diff
@@ -16,6 +16,7 @@ dependencies {
     commonMainApi(projects.units)
 
     "androidMainImplementation"(libs.recyclerView)
+    "androidMainImplementation"(projects.unitsDatabinding)
```
Comment (Alex009) [inline sample/mpp-library/build.gradle.kts:19]: no, we should extract databinding logic from mpp-library at all. also we should remove kapt plugin from mpp and remove javaxAnnotation from mpp

File: units-databinding/src/main/kotlin/dev.icerock.moko.units.databinding/BindingAdapters.kt [hunk 11]
```diff
@@ -95,3 +99,40 @@ private fun getLifecycleOwnerFromContext(context: Context): LifecycleOwner {
     if (context is ContextWrapper) return getLifecycleOwnerFromContext(context.baseContext)
     throw IllegalArgumentException("context $context not implement LifecycleOwner")
 }
+
+fun RecyclerView.bindUnits(lifecycleOwner: LifecycleOwner, units: LiveData<List<UnitItem>>) {
```
Comment (Alex009) [inline units-databinding/src/main/kotlin/dev.icerock.moko.units.databinding/BindingAdapters.kt:103]: it's not databinding adapter. it's seems like it for viewbinding

File: units-databinding/src/main/kotlin/dev.icerock.moko.units.databinding/BindingAdapters.kt [hunk 12]
```diff
@@ -95,3 +99,40 @@ private fun getLifecycleOwnerFromContext(context: Context): LifecycleOwner {
...
+            unitsAdapter.units = it.orEmpty()
+        }
+    }
+}
+
+fun RecyclerView.bindPagingUnits(
+    lifecycleOwner: LifecycleOwner,
+    units: LiveData<List<TableUnitItem>?>,
+    nextPageAction: () -> Unit
+) {
```
Comment (Alex009) [inline units-databinding/src/main/kotlin/dev.icerock.moko.units.databinding/BindingAdapters.kt:118]: it's not databinding adapter. it's seems like it for viewbinding

File: units/build.gradle.kts [hunk 13]
```diff
@@ -10,9 +10,6 @@ plugins {
     id("detekt-convention")
 }
 
-group = "dev.icerock.moko"
-version = libs.versions.mokoUnitsVersion.get()
-
 android.buildFeatures.dataBinding = true
```
Comment (Alex009) [inline units/build.gradle.kts:13]: databinding should be removed here. and kapt too, and all from databinding

**Review by Alex009 — CHANGES_REQUESTED** [review]

File: sample/mpp-library/build.gradle.kts (hunk 10 above)
Comment (anton6tak) [reply sample/mpp-library/build.gradle.kts:19]: removed kapt, javaxAnnotation, and databinding enabled from mpp

File: units-databinding/src/main/kotlin/dev.icerock.moko.units.databinding/BindingAdapters.kt (hunk 11 above)
Comment (anton6tak) [reply units-databinding/src/main/kotlin/dev.icerock.moko.units.databinding/BindingAdapters.kt:103]: moved to units-viewbinding/.../RecyclerViewBinding.kt

File: units-databinding/src/main/kotlin/dev.icerock.moko.units.databinding/BindingAdapters.kt (hunk 12 above)
Comment (anton6tak) [reply units-databinding/src/main/kotlin/dev.icerock.moko.units.databinding/BindingAdapters.kt:118]: moved to units-viewbinding/.../RecyclerViewBinding.kt

File: units/build.gradle.kts (hunk 13 above)
Comment (anton6tak) [reply units/build.gradle.kts]: removed

File: build.gradle.kts (hunk 2 above)
Comment (anton6tak) [reply build.gradle.kts:6]: to add api-generated

File: README.md [hunk 14]
```diff
@@ -64,14 +62,27 @@ multiplatformUnits {
...
 [Here](sample/ios-app/Podfile)'s an example.
 To simplify integration with MultiPlatformFramework you can use [mobile-multiplatform-plugin](https://github.com/icerockdev/mobile-multiplatform-gradle-plugin)**.  
 `MultiPlatformLibraryUnits` CocoaPod contains an swift additions for `UnitDataSource`s of `UITableView`/`UICollectionView`.
 
+In android build.gradle
+
+databinding usage
+```
+apply plugin: "dev.icerock.mobile.multiplatform-units"
+implementation(projects.unitsDatabinding)
```
Comment (Alex009) [inline README.md:78]: plugin configuration should be only in databinding case

```
multiplatformUnits {
    classesPackage = "org.example.library.units"
    dataBindingPackage = "org.example.library"
    layoutsSourceSet = "main"
}
```

Comment (Alex009) [reply README.md:78]: library artifact should be as string

File: README.md [hunk 15]
```diff
@@ -64,14 +62,27 @@ multiplatformUnits {
...
+```
+
+viewbinding usage 
+```
+implementation(projects.unitsViewbinding)
```
Comment (Alex009) [inline README.md:83]: library artifact should be as string

**Review by Alex009 — CHANGES_REQUESTED** [review]

**Review by Alex009 — APPROVED** [review]

**Pull request merged**